libshvproto-macros/src/lib.rs

```diff
@@ -80,11 +128,88 @@ pub fn derive_from_rpcvalue(item: TokenStream) -> TokenStream {
 
                 impl From<#struct_identifier> for shvproto::RpcValue {
                     fn from(value: #struct_identifier) -> Self {
-                        [#map_initializers].into_iter().collect::<shvproto::rpcvalue::Map>().into()
+                        let mut map = shvproto::rpcvalue::Map::new();
+                        #rpcvalue_inserts
+                        map.into()
+                    }
+                }
+            }
+        },
+        syn::Data::Enum(syn::DataEnum { variants, .. }) => {
+            let mut try_froms = quote!{};
+            let mut match_arms = quote!{};
+            let mut expected_types = quote!{""};
+            for variant in variants {
+                let variant_ident = &variant.ident;
+                let mut add_type_matcher = |should, dest_variant_type, block| {
+                    if should {
+                        expected_types.extend(quote! {+ " " + stringify!(#dest_variant_type)});
+
+                        try_froms.extend(quote!{
+                            if let shvproto::Value::#dest_variant_type = value.value() {
```

Comment:
I'd consider generating `match` arms:

``` rust
try_froms.extend(quote!{
    shvproto::Value::#dest_variant_type => Ok(<#struct_identifier>::#variant_ident #block),
});
```

It could also generate better code than a sequence of single `if let` statements without else.